zebra-consensus/src/chain.rs

```diff
@@ -139,196 +77,74 @@ where
-                // Log a message on unexpected out of order blocks.
-                //
-                // The sync service rejects most of these blocks, but we
-                // still want to know if a large number get through.
-                if is_unexpected_gap {
-                    tracing::debug!("large block height gap: this block or the previous block is out of order");
-                }
-
-                let verified_hash = block_verifier
-                    .ready_and()
-                    .await?
-                    .call(block.clone())
-                    .await?;
-                assert_eq!(verified_hash, hash, "block verifier returned wrong hash: hashes must be equal");
-            } else {
-                let verified_hash = checkpoint_verifier
-                    .expect("missing checkpoint verifier: verifier must be Some if max checkpoint height is Some")
-                    .ready_and()
-                    .await?
-                    .call(block.clone())
-                    .await?;
-                assert_eq!(verified_hash, hash, "checkpoint verifier returned wrong hash: hashes must be equal");
-            }
-
-            tracing::trace!(?height, ?hash, "verified block");
-            metrics::gauge!(
-                "chain.verified.block.height",
-                height.expect("valid blocks have a block height").0 as _
+        // Log a message on unexpected out of order blocks.
+        //
+        // The sync service rejects most of these blocks, but we
+        // still want to know if a large number get through.
+        if is_unexpected_gap {
+            tracing::debug!(
+                "large block height gap: this block or the previous block is out of order"
             );
-            metrics::counter!("chain.verified.block.count", 1);
-
-            Ok(hash)
         }
-        .instrument(span)
-        .boxed()
-    }
-}
 
-/// Returns true if block_height is higher than the maximum checkpoint
-/// height. Also returns true if there is no maximum checkpoint height.
-///
-/// Returns false if the block does not have a height.
-fn is_higher_than_max_checkpoint(
-    block_height: Option<block::Height>,
-    max_checkpoint_height: Option<block::Height>,
-) -> bool {
-    match (block_height, max_checkpoint_height) {
-        (Some(block_height), Some(max_checkpoint_height)) => block_height > max_checkpoint_height,
-        (_, None) => true,
-        (None, _) => false,
+        self.last_block_height = height;
+
+        // The only valid block without a coinbase height is the genesis block,
+        // which omitted it by mistake.  So for the purposes of routing requests,
+        // we can interpret a missing coinbase height as 0; the checkpoint verifier
+        // will reject it.
+        if height.unwrap_or(block::Height(0)) < self.max_checkpoint_height {
+            self.checkpoint.call(block)
+        } else {
+            self.block.call(block)
```

Comment:
Looks like these are missing a `ready_and` before the `call`.

Reply:
Good point, we should be calling `poll_ready` on `checkpoint` and `block` in our `poll_ready`.

Reply:
Fixed in 47de2157